// @applitools/ufg-client/dist/utils/to-render-result.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.toRenderResult = void 0;
function toRenderResult(rawResult, { renderId }) {
    var _a;
    return {
        renderId,
        status: rawResult.status,
        error: rawResult.error,
        image: rawResult.imageLocation,
        dom: rawResult.domLocation,
        locationInViewport: rawResult.imagePositionInActiveFrame,
        locationInView: rawResult.imagePositionInActiveFrame,
        fullViewSize: rawResult.fullPageSize,
        selectorRegions: (_a = rawResult.selectorRegions) === null || _a === void 0 ? void 0 : _a.map((regions) => {
            return regions === null || regions === void 0 ? void 0 : regions.map(region => ({
                ...region,
                x: Math.max(0, region.x - rawResult.imagePositionInActiveFrame.x),
                y: Math.max(0, region.y - rawResult.imagePositionInActiveFrame.y),
            }));
        }),
    };
}
exports.toRenderResult = toRenderResult;
