// @applitools/core/dist/autonomous/create-render-results.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.makeCreateRenderResults = void 0;
const ufg_client_1 = require("@applitools/ufg-client");
const to_base_check_settings_1 = require("../automation/utils/to-base-check-settings");
function makeCreateRenderResults(_options) {
    return async function createRenderResults({ renders, rawResults, }) {
        const results = rawResults.map((rawResult, index) => (0, ufg_client_1.toRenderResult)(rawResult, { renderId: renders[index].render.renderId }));
        return results.map(({ selectorRegions, renderId, status, error, ...baseTarget }, index) => {
            const { elementReferences: selectors, getBaseCheckSettings } = (0, to_base_check_settings_1.toBaseCheckSettings)({
                settings: renders[index].settings,
            });
            const baseSettings = getBaseCheckSettings({
                calculatedRegions: selectors.calculate.map((_, index) => {
                    var _a;
                    return ({
                        regions: (_a = selectorRegions === null || selectorRegions === void 0 ? void 0 : selectorRegions[index]) !== null && _a !== void 0 ? _a : [],
                    });
                }),
            });
            baseSettings.renderId = renderId;
            return {
                status,
                error,
                target: { ...baseTarget, isTransformed: true },
                settings: baseSettings,
                environment: renders[index].environment,
                account: renders[index].account,
            };
        });
    };
}
exports.makeCreateRenderResults = makeCreateRenderResults;
