// @applitools/core/dist/autonomous/create-render-target.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.makeCreateRenderTarget = void 0;
function makeCreateRenderTarget({ core, logger: defaultLogger }) {
    return async function createRenderTarget({ snapshot, logger = defaultLogger, }) {
        const ufgClient = await core.getUFGClient({
            settings: {
                ...snapshot.account.ufgServer,
                eyesServerUrl: snapshot.account.eyesServer.eyesServerUrl,
                apiKey: snapshot.account.eyesServer.apiKey,
            },
            logger,
        });
        const target = await ufgClient.createRenderTarget({
            snapshot: snapshot.target,
            settings: {
                environment: snapshot.environment,
                cookies: snapshot.target.cookies,
                headers: {
                    Referer: snapshot.url,
                    'User-Agent': snapshot.userAgent,
                    ...snapshot.settings.headers,
                },
                proxy: snapshot.account.eyesServer.proxy,
                autProxy: snapshot.settings.autProxy,
            },
            logger,
        });
        return { ...snapshot, target };
    };
}
exports.makeCreateRenderTarget = makeCreateRenderTarget;
